app/org/sagebionetworks/bridge/services/AppConfigService.java

```diff
@@ -87,9 +100,23 @@ public AppConfig getAppConfigForUser(CriteriaContext context, boolean throwExcep
         } else if (matches.size() != 1) {
             // If there is more than one match, return the one created first, but log an error
             LOG.error("CriteriaContext matches more than one app config: criteriaContext=" + context + ", appConfigs="+matches);
-            return matches.get(0);
         }
-        return matches.get(0);
+        AppConfig matched = matches.get(0);
+        // Resolve survey references to pick up survey identifiers
+        matched.setSurveyReferences(matched.getSurveyReferences().stream()
+            .map(surveyReference -> resolveSurvey(surveyReference)).collect(Collectors.toList()));
+        return matched;
+    }
+
+    SurveyReference resolveSurvey(SurveyReference surveyRef) {
+        GuidCreatedOnVersionHolder surveyKeys = new GuidCreatedOnVersionHolderImpl(surveyRef);
+        // Survey references can be bogus, and should be return unmodified if they don't reference an existing survey
+        try {
+            Survey survey = surveyService.getSurvey(surveyKeys);
+            return new SurveyReference(survey.getIdentifier(), survey.getGuid(), new DateTime(survey.getCreatedOn()));
```

Comment:
Three optimizations:

1. Can we create a flag in SurveyService to not load SurveyElements? Otherwise, we end up loading a bunch of survey elements that we just throw away.
2. getSurvey() requires that guid and createdOn exist, so this method is really only used to get the identifier. Can we add a check to skip loading the survey if the identifier is already present?
3. Lastly, a SurveyReference without a createdOn is a perfectly valid SurveyReference, and this will crash this code with an IllegalArgumentExcecption. How do we handle those cases?

Reply:
In the AppConfigs, we currently require all the schema and survey references to be "hard" references... that is, they reference an explicit version or createdOn timestamp (soft references are not allowed). I could check for that here in case we change this later, but right now #3 won't be an issue.

I will add the other two optimizations.

Reply:
We don't want the app configs to cache a survey identifier value that is later updated in the survey itself, so either we're going to remove the identifiers in the app config before we persist it, which will cause an update of the identifiers on every retrieval, OR we look up the identifiers on every retrieval and ignore any values in the record. Either way, we're looking these up on each request, however, app configs are cached at the controller tier.

Reply:
There's a lot to this, but I'll check the parts to make sure this is all true. Then as you say, it's save to save the identifier in the app config so we can do that and then skip looking it up again.